File: packages/android/locanara/src/main/kotlin/com/locanara/rag/DocumentChunker.kt
package com.locanara.rag

import java.text.BreakIterator
import java.util.Locale
import java.util.UUID

/**
 * Configuration for document chunking
 */
data class ChunkingConfig(
    /** Target chunk size in characters (approximate) */
    val targetChunkSize: Int = 512,
    /** Overlap between chunks in characters */
    val chunkOverlap: Int = 50,
    /** Whether to respect sentence boundaries */
    val respectSentences: Boolean = true,
    /** Minimum chunk size (smaller chunks will be merged) */
    val minChunkSize: Int = 100
) {
    companion object {
        /** Default configuration optimized for on-device embedding */
        val DEFAULT = ChunkingConfig()

        /** Configuration for longer documents */
        val LONG_DOCUMENT = ChunkingConfig(
            targetChunkSize = 1024,
            chunkOverlap = 100,
            respectSentences = true,
            minChunkSize = 200
        )
    }
}

/**
 * A chunk of text extracted from a document
 */
data class DocumentChunk(
    /** Unique identifier for this chunk */
    val id: String = UUID.randomUUID().toString(),
    /** The text content of this chunk */
    val content: String,
    /** Position of this chunk in the original document (0-indexed) */
    val index: Int,
    /** Character offset in the original document */
    val startOffset: Int,
    /** Character end offset in the original document */
    val endOffset: Int,
    /** Optional metadata */
    val metadata: Map<String, String>? = null
)

/**
 * Statistics about chunking results
 */
data class ChunkingStats(
    val count: Int,
    val minSize: Int,
    val maxSize: Int,
    val avgSize: Int,
    val totalSize: Int
)

/**
 * Handles splitting documents into chunks for RAG indexing
 */
class DocumentChunker(
    private val config: ChunkingConfig = ChunkingConfig.DEFAULT
) {
    /**
     * Chunk a document into smaller pieces for embedding
     * @param text The full document text
     * @param metadata Optional metadata to attach to each chunk
     * @return Array of document chunks
     */
    fun chunk(
        text: String,
        metadata: Map<String, String>? = null
    ): List<DocumentChunk> {
        if (text.isEmpty()) return emptyList()

        return if (config.respectSentences) {
            chunkBySentences(text, metadata)
        } else {
            chunkByCharacters(text, metadata)
        }
    }

    /**
     * Chunk text respecting sentence boundaries
     */
    private fun chunkBySentences(
        text: String,
        metadata: Map<String, String>?
    ): List<DocumentChunk> {
        val sentences = splitIntoSentences(text)

        if (sentences.isEmpty()) {
            return chunkByCharacters(text, metadata)
        }

        val chunks = mutableListOf<DocumentChunk>()
        var currentChunkText = StringBuilder()
        var chunkStartOffset = 0
        var currentOffset = 0

        for (sentence in sentences) {
            val potentialChunk = currentChunkText.toString() + sentence

            if (potentialChunk.length > config.targetChunkSize && currentChunkText.isNotEmpty()) {
                // Current chunk is full, save it
                val chunkContent = currentChunkText.toString().trim()
                if (chunkContent.length >= config.minChunkSize) {
                    chunks.add(
                        DocumentChunk(
                            content = chunkContent,
                            index = chunks.size,
                            startOffset = chunkStartOffset,
                            endOffset = chunkStartOffset + currentChunkText.length,
                            metadata = metadata
                        )
                    )
                }

                // Start new chunk with overlap
                val overlapText = getOverlapText(currentChunkText.toString())
                currentChunkText = StringBuilder(overlapText + sentence)
                chunkStartOffset = currentOffset - overlapText.length
            } else {
                if (currentChunkText.isEmpty()) {
                    chunkStartOffset = currentOffset
                }
                currentChunkText.append(sentence)
            }
            currentOffset += sentence.length
        }

        // Add remaining text as final chunk
        if (currentChunkText.isNotEmpty()) {
            val chunkContent = currentChunkText.toString().trim()
            if (chunkContent.length < config.minChunkSize && chunks.isNotEmpty()) {
                // Merge with previous chunk
                val lastChunk = chunks.removeAt(chunks.size - 1)
                val mergedContent = lastChunk.content + " " + chunkContent
                chunks.add(
                    lastChunk.copy(
                        content = mergedContent,
                        endOffset = chunkStartOffset + currentChunkText.length
                    )
                )
            } else if (chunkContent.length >= config.minChunkSize) {
                chunks.add(
                    DocumentChunk(
                        content = chunkContent,
                        index = chunks.size,
                        startOffset = chunkStartOffset,
                        endOffset = minOf(chunkStartOffset + currentChunkText.length, text.length),
                        metadata = metadata
                    )
                )
            }
        }

        return chunks
    }

    /**
     * Chunk text by character count (simpler approach)
     */
    private fun chunkByCharacters(
        text: String,
        metadata: Map<String, String>?
    ): List<DocumentChunk> {
        val chunks = mutableListOf<DocumentChunk>()
        var currentIndex = 0

        while (currentIndex < text.length) {
            val chunkSize = minOf(config.targetChunkSize, text.length - currentIndex)
            val chunkText = text.substring(currentIndex, currentIndex + chunkSize).trim()

            if (chunkText.length >= config.minChunkSize) {
                chunks.add(
                    DocumentChunk(
                        content = chunkText,
                        index = chunks.size,
                        startOffset = currentIndex,
                        endOffset = currentIndex + chunkSize,
                        metadata = metadata
                    )
                )
            }

            // Move forward with overlap
            val moveDistance = maxOf(1, chunkSize - config.chunkOverlap)
            currentIndex += moveDistance
        }

        return chunks
    }

    /**
     * Split text into sentences using BreakIterator
     */
    private fun splitIntoSentences(text: String): List<String> {
        val sentences = mutableListOf<String>()
        val iterator = BreakIterator.getSentenceInstance(Locale.getDefault())
        iterator.setText(text)

        var start = iterator.first()
        var end = iterator.next()

        while (end != BreakIterator.DONE) {
            sentences.add(text.substring(start, end))
            start = end
            end = iterator.next()
        }

        return sentences
    }

    /**
     * Get overlap text from the end of a chunk
     */
    private fun getOverlapText(text: String): String {
        if (config.chunkOverlap <= 0) return ""
        val overlapSize = minOf(config.chunkOverlap, text.length)
        return text.substring(text.length - overlapSize)
    }

    /**
     * Estimate the number of chunks for a given text
     */
    fun estimateChunkCount(text: String): Int {
        if (text.isEmpty()) return 0
        val effectiveChunkSize = config.targetChunkSize - config.chunkOverlap
        return maxOf(1, (text.length + effectiveChunkSize - 1) / effectiveChunkSize)
    }

    /**
     * Get chunking statistics for debugging
     */
    fun getChunkingStats(chunks: List<DocumentChunk>): ChunkingStats {
        if (chunks.isEmpty()) {
            return ChunkingStats(count = 0, minSize = 0, maxSize = 0, avgSize = 0, totalSize = 0)
        }

        val sizes = chunks.map { it.content.length }
        val total = sizes.sum()

        return ChunkingStats(
            count = chunks.size,
            minSize = sizes.minOrNull() ?: 0,
            maxSize = sizes.maxOrNull() ?: 0,
            avgSize = total / chunks.size,
            totalSize = total
        )
    }
}

File: packages/apple/Sources/RAG/DocumentChunker.swift
import Foundation
import NaturalLanguage

/// Configuration for document chunking
public struct ChunkingConfig: Sendable {
    /// Target chunk size in characters (approximate)
    public let targetChunkSize: Int

    /// Overlap between chunks in characters
    public let chunkOverlap: Int

    /// Whether to respect sentence boundaries
    public let respectSentences: Bool

    /// Minimum chunk size (smaller chunks will be merged)
    public let minChunkSize: Int

    public init(
        targetChunkSize: Int = 512,
        chunkOverlap: Int = 50,
        respectSentences: Bool = true,
        minChunkSize: Int = 100
    ) {
        self.targetChunkSize = targetChunkSize
        self.chunkOverlap = chunkOverlap
        self.respectSentences = respectSentences
        self.minChunkSize = minChunkSize
    }

    /// Default configuration optimized for on-device embedding
    public static let `default` = ChunkingConfig()

    /// Configuration for longer documents
    public static let longDocument = ChunkingConfig(
        targetChunkSize: 1024,
        chunkOverlap: 100,
        respectSentences: true,
        minChunkSize: 200
    )
}

/// A chunk of text extracted from a document
public struct DocumentChunk: Sendable, Codable {
    /// Unique identifier for this chunk
    public let id: String

    /// The text content of this chunk
    public let content: String

    /// Position of this chunk in the original document (0-indexed)
    public let index: Int

    /// Character offset in the original document
    public let startOffset: Int

    /// Character end offset in the original document
    public let endOffset: Int

    /// Optional metadata
    public let metadata: [String: String]?

    public init(
        id: String = UUID().uuidString,
        content: String,
        index: Int,
        startOffset: Int,
        endOffset: Int,
        metadata: [String: String]? = nil
    ) {
        self.id = id
        self.content = content
        self.index = index
        self.startOffset = startOffset
        self.endOffset = endOffset
        self.metadata = metadata
    }
}

/// Handles splitting documents into chunks for RAG indexing
public actor DocumentChunker {
    private let config: ChunkingConfig
    private let tokenizer: NLTokenizer

    public init(config: ChunkingConfig = .default) {
        self.config = config
        self.tokenizer = NLTokenizer(unit: .sentence)
    }

    /// Chunk a document into smaller pieces for embedding
    /// - Parameters:
    ///   - text: The full document text
    ///   - metadata: Optional metadata to attach to each chunk
    /// - Returns: Array of document chunks
    public func chunk(
        text: String,
        metadata: [String: String]? = nil
    ) -> [DocumentChunk] {
        guard !text.isEmpty else { return [] }

        if config.respectSentences {
            return chunkBySentences(text: text, metadata: metadata)
        } else {
            return chunkByCharacters(text: text, metadata: metadata)
        }
    }

    /// Chunk text respecting sentence boundaries
    private func chunkBySentences(
        text: String,
        metadata: [String: String]?
    ) -> [DocumentChunk] {
        tokenizer.string = text

        var sentences: [(String, Range<String.Index>)] = []
        tokenizer.enumerateTokens(in: text.startIndex..<text.endIndex) { range, _ in
            let sentence = String(text[range])
            sentences.append((sentence, range))
            return true
        }

        guard !sentences.isEmpty else {
            // Fallback to character-based chunking
            return chunkByCharacters(text: text, metadata: metadata)
        }

        var chunks: [DocumentChunk] = []
        var currentChunkText = ""
        var chunkStartIndex: String.Index = text.startIndex

        for (_, (sentence, range)) in sentences.enumerated() {
            let potentialChunk = currentChunkText + sentence

            if potentialChunk.count > config.targetChunkSize && !currentChunkText.isEmpty {
                // Current chunk is full, save it
                let startOffset = text.distance(from: text.startIndex, to: chunkStartIndex)
                let endOffset = startOffset + currentChunkText.count

                let chunk = DocumentChunk(
                    content: currentChunkText.trimmingCharacters(in: .whitespacesAndNewlines),
                    index: chunks.count,
                    startOffset: startOffset,
                    endOffset: endOffset,
                    metadata: metadata
                )

                if chunk.content.count >= config.minChunkSize {
                    chunks.append(chunk)
                }

                // Start new chunk with overlap
                let overlapText = getOverlapText(from: currentChunkText)
                currentChunkText = overlapText + sentence
                chunkStartIndex = range.lowerBound

                // Adjust for overlap
                if !overlapText.isEmpty {
                    let overlapStart = text.index(range.lowerBound, offsetBy: -overlapText.count, limitedBy: text.startIndex) ?? text.startIndex
                    chunkStartIndex = overlapStart
                }
            } else {
                currentChunkText = potentialChunk
                if chunks.isEmpty && currentChunkText == sentence {
                    chunkStartIndex = range.lowerBound
                }
            }
        }

        // Add remaining text as final chunk
        if !currentChunkText.isEmpty {
            let startOffset = text.distance(from: text.startIndex, to: chunkStartIndex)
            let endOffset = startOffset + currentChunkText.count

            let chunk = DocumentChunk(
                content: currentChunkText.trimmingCharacters(in: .whitespacesAndNewlines),
                index: chunks.count,
                startOffset: startOffset,
                endOffset: min(endOffset, text.count),
                metadata: metadata
            )

            // Merge with previous chunk if too small
            if chunk.content.count < config.minChunkSize && !chunks.isEmpty {
                let lastChunk = chunks.removeLast()
                let mergedContent = lastChunk.content + " " + chunk.content
                chunks.append(DocumentChunk(
                    id: lastChunk.id,
                    content: mergedContent,
                    index: lastChunk.index,
                    startOffset: lastChunk.startOffset,
                    endOffset: chunk.endOffset,
                    metadata: metadata
                ))
            } else if chunk.content.count >= config.minChunkSize {
                chunks.append(chunk)
            } else if chunks.isEmpty && !chunk.content.isEmpty {
                // Always create at least one chunk even if smaller than minChunkSize
                chunks.append(chunk)
            }
        }

        return chunks
    }

    /// Chunk text by character count (simpler approach)
    private func chunkByCharacters(
        text: String,
        metadata: [String: String]?
    ) -> [DocumentChunk] {
        var chunks: [DocumentChunk] = []
        var currentIndex = text.startIndex
        var chunkIndex = 0

        while currentIndex < text.endIndex {
            let remainingDistance = text.distance(from: currentIndex, to: text.endIndex)
            let chunkSize = min(config.targetChunkSize, remainingDistance)

            let endIndex = text.index(currentIndex, offsetBy: chunkSize)
            let chunkText = String(text[currentIndex..<endIndex])

            let startOffset = text.distance(from: text.startIndex, to: currentIndex)
            let endOffset = startOffset + chunkText.count

            let chunk = DocumentChunk(
                content: chunkText.trimmingCharacters(in: .whitespacesAndNewlines),
                index: chunkIndex,
                startOffset: startOffset,
                endOffset: endOffset,
                metadata: metadata
            )

            // Always add if large enough, or if it's the only chunk (allow short documents)
            if chunk.content.count >= config.minChunkSize || (chunks.isEmpty && !chunk.content.isEmpty) {
                chunks.append(chunk)
                chunkIndex += 1
            }

            // Move forward with overlap
            let moveDistance = max(1, chunkSize - config.chunkOverlap)
            if let nextIndex = text.index(currentIndex, offsetBy: moveDistance, limitedBy: text.endIndex) {
                currentIndex = nextIndex
            } else {
                break
            }
        }

        return chunks
    }

    /// Get overlap text from the end of a chunk
    private func getOverlapText(from text: String) -> String {
        guard config.chunkOverlap > 0 else { return "" }

        let overlapSize = min(config.chunkOverlap, text.count)
        let startIndex = text.index(text.endIndex, offsetBy: -overlapSize)
        return String(text[startIndex...])
    }
}

// MARK: - Convenience Methods

extension DocumentChunker {
    /// Estimate the number of chunks for a given text
    public func estimateChunkCount(for text: String) -> Int {
        guard !text.isEmpty else { return 0 }
        let effectiveChunkSize = config.targetChunkSize - config.chunkOverlap
        return max(1, (text.count + effectiveChunkSize - 1) / effectiveChunkSize)
    }

    /// Get chunking statistics for debugging
    public func getChunkingStats(chunks: [DocumentChunk]) -> ChunkingStats {
        guard !chunks.isEmpty else {
            return ChunkingStats(count: 0, minSize: 0, maxSize: 0, avgSize: 0, totalSize: 0)
        }

        let sizes = chunks.map { $0.content.count }
        let total = sizes.reduce(0, +)

        return ChunkingStats(
            count: chunks.count,
            minSize: sizes.min() ?? 0,
            maxSize: sizes.max() ?? 0,
            avgSize: total / chunks.count,
            totalSize: total
        )
    }
}

/// Statistics about chunking results
public struct ChunkingStats: Sendable {
    public let count: Int
    public let minSize: Int
    public let maxSize: Int
    public let avgSize: Int
    public let totalSize: Int
}
